drivers/spi/spi_xec_gpspi.c

```diff
@@ -0,0 +1,791 @@
+ *    If RX buffer exists write data to it.
+ * TODO Logic to disable these interrupts
+ */
+static int xec_gpspi_xfr_async(const struct device *dev,
+			       const struct spi_config *spi_conf)
+{
+	const struct spi_xec_gpspi_config *cfg = dev->config;
+	struct spi_xec_gpspi_data * const data = dev->data;
+	struct xec_gpspi_regs * const regs = cfg->regs;
+	struct spi_context *ctx = &data->ctx;
+	uint8_t txb;
+
+	if ((regs->status & XEC_GPSPI_STS_RXBF_TXBE) != BIT(XEC_GPSPI_STATUS_TXBE_POS)) {
+		return -EBUSY;
+	}
+
+	if (spi_context_tx_buf_on(ctx) || spi_context_rx_buf_on(ctx)) {
+		data->ctx_longest_buf_len = spi_context_longest_current_buf(ctx);
+		data->isr_ctx_done = 0u;
+
+		/* Enable GPSPI controller's RXBF interrupt.
+		 * Controller has no interrupt enables. We turn it
+		 * on in the GIRQ.
+		 */
+		mchp_xec_ecia_girq_src_en(cfg->irqrx.girq, cfg->irqrx.girq_pos);
+
+		if (spi_context_tx_buf_on(ctx)) {
+			txb = *ctx->tx_buf;
+			spi_context_update_tx(ctx, 1, 1);
+			regs->control |= BIT(XEC_GPSPI_CTRL_BI_DIR_OUT_EN_POS);
+			regs->tx_data = txb;
+		} else {
+			regs->control &= ~BIT(XEC_GPSPI_CTRL_BI_DIR_OUT_EN_POS);
+			regs->tx_data = 0;
+		}
+	} else {
+		spi_context_complete(ctx, dev, 0);
+	}
+
+	return 0;
+}
+#endif
+
+/* SPI model is for every clock edge where data is transmitted we must
+ * read data in on the sample clock edge. The chosen SPI mode specifies
+ * the transmit and input sample clock edges.
+ * If no RX buffer corresponding to a TX buffer is provided the sampled
+ * data is discarded. The XEC GPSPI controller requires sampled data to
+ * be always read or its overrun status gets stuck requiring multiple
+ * reads of the RX data register to clear. The controller also has an
+ * auto-read HW feature where reads of the RX data register cause the
+ * controller to generate clocks for the next byte. We will not use
+ * auto-read for synchronous transfer.
+ */
+static int xec_gpspi_xfr(const struct device *dev,
+			 const struct spi_config *spi_conf,
+			 const struct spi_buf_set *tx_bufs,
+			 const struct spi_buf_set *rx_bufs,
+			 bool asynchronous,
+			 spi_callback_t cb,
+			 void *userdata)
+{
+	const struct spi_xec_gpspi_config *cfg = dev->config;
+	struct xec_gpspi_regs * const regs = cfg->regs;
+	struct spi_xec_gpspi_data * const data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	int ret = 0;
+
+	spi_context_lock(ctx, asynchronous, cb, userdata, spi_conf);
+
+	ret = gpspi_configure(dev, spi_conf);
+	if (ret != 0) {
+		spi_context_release(ctx, ret);
+		return ret;
+	}
+
+	spi_context_cs_control(&data->ctx, true);
+	regs->control |= BIT(XEC_GPSPI_CTRL_CE_POS);
+
+	spi_context_buffers_setup(ctx, tx_bufs, rx_bufs, 1);
+
+#ifdef CONFIG_SPI_ASYNC
+	if (asynchronous) {
+		ret = xec_gpspi_xfr_async(dev, spi_conf);
+		if (ret) {
+			regs->control &= ~(BIT(XEC_GPSPI_CTRL_CE_POS));
+			spi_context_unlock_unconditionally(&data->ctx);
+		}
+		return ret;
+	}
+#endif
+	ret = xec_gpspi_xfr_sync(dev, spi_conf);
```

Comment:
See above comment. Your driver does not need it as it can run on interrupt mode.
CONFIG_SPI_ASYNC is only about being able to get the cb/userdata handled relevantly, not about driver's internal logic.